File: daily/572-Subtree-of-another-tree.py
class Solution:
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        def dfs(root, subRoot):
            if not root or not subRoot:
                return False
            if isSameTree(root, subRoot):
                return True
            return dfs(root.left, subRoot) or dfs(root.right, subRoot)

        def isSameTree(node1, node2):
            if not node1 and not node2:
                return True
            if not node1 or not node2:
                return False
            return (node1.val == node2.val and isSameTree(node1.left, node2.left) and isSameTree(node1.right, node2.right))
        
        return dfs(root, subRoot)

File: daily/Binary Tree/572-Subtree-of-another-tree.go
package daily

func isSubtree(root *TreeNode, subRoot *TreeNode) bool {
    if root == nil || subRoot == nil {
        return false
    }
    if isSameTree(root, subRoot) {
        return true
    }
    return isSubtree(root.Left, subRoot) || isSubtree(root.Right, subRoot)
}

func isSameTree(node1 *TreeNode, node2 *TreeNode) bool {
    if node1 == nil && node2 == nil {
        return true
    }
    if node1 == nil || node2 == nil {
        return false
    }
    if node1.Val != node2.Val {
        return false
    }
    left := isSameTree(node1.Left, node2.Left)
    right := isSameTree(node1.Right, node2.Right)
    return left && right
}